### PR TITLE
Votes

### PR DESCRIPTION
**Pull request by mohelsaka**
in this branch we added:
- Votes for Questions, Answers and Videos.
- Reputation for the user displayed besides his name in the top bar.
- Authorization for performing some actions e.g. voting up and down.
- Simple popup message to display error messages from authorization.

Note:
There is another branch contains some new updates for profile. this branch called profile_cont.

Regards,

### CODE REVIEW

File: app/controllers/videos_controller.rb [hunk 1]
```diff
@@ -86,13 +87,56 @@ def destroy
...
     @answer = Answer.new
   end
   
   def add_answer
-    @question = Question.find(params[:id])
-    @answer = @question.answers.create(:body => params[:answer][:body], :user_id => current_user.id)
+    @question = Question.find params[:id]
+    @answer = Answer.new(:body => params[:answer][:body], :user_id => current_user.id, :question_id => params[:id])
+    
+    params[:error_place] = "submit-ans-#{params[:id]}"
```
Comment (mohelsaka) [inline app/controllers/videos_controller.rb]: why params however, it can be instance variable

File: app/controllers/videos_controller.rb [hunk 2]
```diff
@@ -86,13 +87,56 @@ def destroy
   end
   
   def add_question
-    @video = Video.find(params[:id])
-    @question = @video.questions.create(:body => params[:question][:body], :user_id => current_user.id)
+    @question = Question.new(:body => params[:question][:body], :user_id => current_user.id, :video_id => params[:id])
+    
+    params[:error_place] = 'submit-question'
```
Comment (mahabdelaziz) [inline app/controllers/videos_controller.rb]: why params, however it can be instance variable

File: app/controllers/videos_controller.rb [hunk 3]
```diff
@@ -86,13 +87,56 @@ def destroy
...
   def add_answer
-    @question = Question.find(params[:id])
-    @answer = @question.answers.create(:body => params[:answer][:body], :user_id => current_user.id)
+    @question = Question.find params[:id]
+    @answer = Answer.new(:body => params[:answer][:body], :user_id => current_user.id, :question_id => params[:id])
+    
+    params[:error_place] = "submit-ans-#{params[:id]}"
+    authorize! :create, @answer, :message => t('.can_not_answer')
+    
+    display_error_popup(params[:error_place], (@answer.errors.messages[:body].join('and'))) unless @answer.save
```
Comment (mahabdelaziz) [inline app/controllers/videos_controller.rb]: refactor this long line in a method

File: app/controllers/videos_controller.rb [hunk 4]
```diff
@@ -86,13 +87,56 @@ def destroy
...
+  end
+  
+  def vote
+    target = params[:target]
+    id = params[:id]
+    
+    object = if target == 'v'
```
Comment (mahabdelaziz) [inline app/controllers/videos_controller.rb]: eval(target.classify).find(id)

File: app/models/ability.rb [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    
+    points = user.reputation_value_for :points
+    # 10 points to add answer
+    if points >= 10
+      can :create, Answer
+    end
+    
+    # user can't vote on his Videos, Answers and Questions
+    can :vote, [Answer, Question, Video] do |object|
+      user.persisted? && object.user_id != user.id
```
Comment (mahabdelaziz) [inline app/models/ability.rb]: no need for persisted

File: app/models/question.rb [hunk 6]
```diff
@@ -4,4 +4,11 @@ class Question < ActiveRecord::Base
   has_many :answers
   
   validates :body, :presence => true
+
+  acts_as_votable
```
Comment (mahabdelaziz) [inline app/models/question.rb]: ya3

File: app/models/answer.rb [hunk 7]
```diff
@@ -3,4 +3,11 @@ class Answer < ActiveRecord::Base
   belongs_to :user
   
   validates :body, :presence => true
+
+  acts_as_votable
```
Comment (mahabdelaziz) [inline app/models/answer.rb]: ya3 and remove from gemfile

File: app/views/videos/_answer.html.haml [hunk 8]
```diff
@@ -1,7 +1,10 @@
 .answer
-  = image_tag 'answer_icon.png', :class => 'icon'  
-  .question-body
-    = answer.body
+  .pull-left
+    = image_tag 'answer_icon.png', :class => 'icon'  
+  .q-data
+    .q-info
+      = link_to answer.user.name, :url => profiles_url(answer.user.id)
```
Comment (mahabdelaziz) [inline app/views/videos/_answer.html.haml]: no :url

**Pull request merged**